### ISSUE TITLE
[DOC] Domain Model

### ISSUE THREAD

**Issue by anitahaurie** [t-vaudry/project_emu#3]
"A domain model, entirely static, showing its principal entities and their relationships. No methods should be included."

Assigning this one to Steve as he will be making the diagram. However, everyone can (and should) post their ideas here as comments.

**Comment by anitahaurie** [comment]
At the very least, we need:

Course:
-Course Number
-Course Name
-Professor
-Time/Date
-Semester
-Prerequisites

Student:
-Name
-ID

**Comment by salaman** [comment]
Some ideas:

### Courses

- subject (SOEN)
- number (341)
- title (Software Process)
- credits (3.0)
- prerequisites: *one-to-many relation: course*
- sections: *one-to-many relation: section*

### Sections

- semester (Winter 2016)
- section code (S)
- type (LEC)
- session (13 week)
- meetings: *one-to-many relation: meetings*

### Meetings

- days of week (W, F)
- start time (8:45 AM)
- end time (10:00 AM)
- location (MB-S2.210)
- instructor

### Students

- name
- student id
- completed courses: _one-to-many relation: course_

--

I have sections and meetings split up because if the start time, end time and days of week are kept inside section, it means a course cannot start on different times on different days (eg. Monday 10AM and Wednesday 2PM). This might not be necessary for SOEN-only courses, but it was for courses in other departments.

Also, if we're assumed to be using only 1 year of data, we can drop the year from the semester column and just do Summer/Winter/Fall. Summer is going to require some extra thinking later for the different sessions within summer.

**Comment by APengue1** [comment]
Since we are required to develop a user sign-in model, it might be useful to add another object to Chris' proposition: Calendar. This way students can save a generated calendar to their profiles.
The Calendar should contain Courses at the very least. 

More detail not relevant to static model for Deliverable 0:
Saved calendars could be saved and loaded to/from an XML file or a means already implemented by Chris, if available. 
One issue is that schedules can change during the semester, especially during the first two weeks (classroom changes, TA's assigned, etc). I have some solutions in mind. We could discuss further if this suggestion is agreed upon.

**Comment by salaman** [comment]
I use a similar class which is just a list of `Section`s (not `Course`s) chosen for a specific generated calendar, so it's similar to what you're saying. It could be included in the domain model.

**Comment by Shmeve** [comment]
So the sample model from the slides has a few extra attributes we haven't considered:
1. GPA attribute: Are we supposed to / do we want to track results of completed courses?
2. User object: Do we just want to add the login information to the student object? Basically combining the Student Record and User from the slide into Student in our model.
 
---
My suggestion would be to add the following objects/relations:
### Course
* all previous attributes
* credits (will be needed if we want to calculate GPAs)

### Student
* all previous attributes
* username
* password
* transcript: one-to-many results

### Result
* grade
* course: one-to-one relation with course

**Comment by Rick--D** [comment]
We can implement a `cumulative GPA` as an attribute for `Student` since Engineering students must maintain a minimum GPA (2.5?) to be considered "in good academic standing"

**Comment by dsecca** [comment]
Here is my Domain Model idea based off of the teacher's diagram and some of the ideas you guys have mentioned. It's rather lengthy but it gives us the big picture of what's going on.

@Shmeve I would say that tracking the results of the previous courses is essential because if a student fails a course, he/she cannot take another course for which the failed one is a prerequisite to. We can use the GPA for this. As for your second point, my "Account" object holds the "Student Record." I think we should merge these two together as you suggested.

**Student**
First name
Last name
ID
_**Student** creates only 1 **Account**_
_**Student** chooses 0 or many **Preferences**_
_**Student** can access 1 **Calendar**_

**Professor**
First Name
Last Name
Program he/she teaches
_**Professor** creates only 1 **Account**_
_**Professor** chooses 0 or many **Preferences**_
_**Professor** can access 1 **Calendar**_

**Account**
First name
Last name
ID
Password
GPA
Program
Co-Op option? (Schedule will be generated differently based off the predetermined course sequence)
_**Account** is an object of only 1 **Schedule Generator**_
_**Account** is tagged to 1 or many **Programs**_ (1 or many since Professors may teach in more than 1 program)
_**Account** is made by many **Professors**_
_**Account** is made by many **Students**_
_**Account** has only 1 **Calendar**_

**Preferences**
Days
Times
_**Preferences** are made by many **Students**_
_**Preferences** are made by many **Professors**_
_**Preferences** are applied by only 1 **Schedule Generator**_

**Program**
Courses
Duration
Course Description
_**Program** is tagged to many **Accounts**_
_**Program** has many **Courses**_
_**Program** registered in 1 **Database**_

**Courses**
Course Code
Course Number
Section
Prerequisites
Semester offered
_**Courses** have 1 or many **Sections**_
_**Courses** are part of one or many **Programs**_
_**Courses** registered in 1 **Database**_

**Section**
Time offered
Classroom
Number of students
Max student capacity (Will be one of the exceptions to be handled)
_**Section** is part of 1 **Course**_
_**Section** is registered in 1 **Database**_

**Database**
Program
Courses
Section
_**Database** has many **Programs**_
_**Database** has many **Courses**_
_**Database** has many **Sections**_
_**Database** is accessed by 1 **Schedule Generator**_

**Schedule Generator**
Preferences
Semester
Account
_**Schedule Generator** accesses only 1 **Database**_
_**Schedule** Generator has many **Accounts**_
_**Schedule** Generator creates many **Calendars**_
_**Schedule** Generator applies many **Preferences**_

**Calendar**
Schedule
_**Calendar** is created by only 1 **Schedule Generator**_
_**Calendar** is in many **Accounts**_
_**Calendar** is accessed by many **Students**_
_**Calendar** is accessed by many **Professors**_

**Comment by salaman** [comment]
We wouldn't need GPA tracking if we decide that the user enters only courses they passed successfully. We would need it if we decide that the user enters their entire academic record including failed/marginal fail courses. I'm not sure what the requirement is, but it doesn't really make much sense for someone to enter failed courses in my opinion.

**Comment by Persepolis1** [comment]
Adding to @dsecca model , we can also add components that constitute the Schedule

**Schedule Items**
Start Time 
End Time
Day Of Week
Course Number (Ex : SOEN 341)
Course Location (Ex : JMSB 2.2)
Many **Schedule Items** in one **Schedule**

**Comment by salaman** [comment]
@dsecca's model is good, however:

- I would personally drop `Professor` from the model and simply have it as an attribute inside `Section` instead of a whole class by itself, since profs have nothing to do with the system (both in login and schedule generation), and merge `Student` and `Account` since they have the same attributes
- "Duration" inside `Program` is unnecessary, since the whole point of this system is to accommodate various durations depending on the student's sequence
- It doesn't make realistic sense to keep track of "Max student capacity" inside `Section`, because that data does not exist for future semesters, and is very hard to get from the SIS (trust me, I tried)
- The `Database` class is not very necessary, since most of the domain model will exist in the database
- Drop "GPA" from `Student`/`Account` because that is dynamically generated based on the student record
- We need a `StudentRecord` or similar combining student, course, and GPA (term taken is irrelevant but we can include that too)
- Your `Calendar` class only consists of a "Schedule", I assume you meant `Sections` there, since a calendar is just a set of sections

**Comment by Shmeve** [comment]
**Student**
First name
Last name
ID
Password
Program
Co-Op
~~_**Student** chooses 0 or many **Preferences**_~~ Student uses generator, relation b/w preferences and generator
_**Student** has 1 **Program**_
_**Student** can access 1 **Calendar**_

**Preferences**
Days
Times
Summers
~~_**Preferences** are made by many **Students**_~~ Student uses generator, relation b/w preferences and generator
_**Preferences** are applied by only 1 **Schedule Generator**_

**Program**
Courses
Course Description
~~_**Program** is tagged to many **Students**_~~ Student has 1 program
_**Program** has many **Courses**_

**Courses**
Course Code
Course Number
Section
Prerequisites
Semester offered
_**Courses** have ~~1~~ 0 or many **Sections**_ (0 sections when course not offered)
~~_**Courses** are part of one or many **Programs**_~~ Redundant

**Section**
Time offered
Classroom
Professor
Number of students
~~_**Section** is part of 1 **Course**_~~ Redundant relation

**Schedule Generator**
Preferences
Semester
Account
~~_**Schedule** Generator has many **Students**_~~ Student uses the generator
_**Schedule** Generator creates many **Calendars**_
_**Schedule** Generator applies many **Preferences**_

**Calendar**
Schedule
~~_**Calendar** is created by only 1 **Schedule Generator**_~~ Redundant
~~_**Calendar** is accessed by many **Students**_~~ Redundant

---
I cleaned up the relations and added the changes suggested by @salaman. If I missed something let me know, otherwise I'll get started on the diagram.

**Comment by anitahaurie** [comment]
I don't think classroom, professor, or number of students are necessary attributes since:
- Classroom is almost never available ahead of time
- Professor is almost never available ahead of time
- Number of students is impossible to know since we aren't handling registration

**Comment by t-vaudry** [comment]
Also, Course description should probably be under Course and not program

**Comment by salaman** [comment]
Both `Program` and `Course` should have a "name" attribute.

**Comment by jhayarevalo** [comment]
Based on deliverable 0, I understood that there were to be multiple types of users. Would it be better to have a "User" class instead of a "Student" one? This class would have an additional "type" attribute (prof, student, anon). If that's too complicated for the relations then maybe we should change our deliverable 0 description for system users.

**Comment by anitahaurie** [comment]
In reference to @jhayarevalo  comment, I believe we should keep the different users in the system description, as we are asked to include them. It would be a good idea however, to add anon and prof to the diagram.

Prof:
ID
password
 -- 1 prof teaches 0..* courses
 -- 1 prof can view 1 calendar

Anon:
 -- 1 Anon can view 1 calendar

**Comment by jhayarevalo** [comment]
Based on deliverable 0, I understood that there were to be multiple types of users using the web tool. Instead of having "Student" as the class, would it be useful to have a "User" class instead with all the attributes already listed in the current "Student" class and add to that a "type" attribute (i.e. prof, student or anon)?

**Comment by Shmeve** [comment]
**Student**
First name
Last name
ID
Password
Program
_**Student** has 1 **Program**_
_**Student** can access 1 **Calendar**_

**Preferences**
Days
Times
Summers
_**Preferences** are applied by only 1 **Schedule Generator**_

**Program**
Name
Courses
Description
_**Program** has many **Courses**_

**Courses**
Course Code
Course Number
Name
Section
Prerequisites
Semester offered
_**Courses** have 0 or many **Sections**_

**Section**
Time offered
Classroom
Professor
Number of students

**Schedule Generator**
Preferences
Semester
Account
_**Schedule** Generator creates many **Calendars**_
_**Schedule** Generator applies many **Preferences**_

**Calendar**
_**Calendar** has many **Courses**_

---
@anitahaurie I think we should keep classroom and professor since they are still attributes of a class, even though they never tell us ahead of time. I cut number of students.

@t-vaudry changed course description to program description, since I am assuming the original intention was to have it describe the program.

@jhayarevalo Where did you see that we require multiple users?
>A brief description of the system, giving its purpose, its functions, and its classes of users without any reference as to how this will be implemented.

The way I understood this was that we just have to specify the types of users and classes we wish to implement. Not that there has to be more than one.

---
I think this should be our final version of the model. If we want to add anonymous users it should be a quick fix anyway. I feel like having accounts for professors would be unnecessary.

**Comment by jhayarevalo** [comment]
@Shmeve it's written under section 1.3 of the Deliverable 0 document (the one that we all added our last name to) but in the actual Project Description given by the teacher, I don't think it's required:
In Project Description
>Students in Software Engineering (and in the whole faculty as well) have to take a certain number of courses to complete their program. 

In Deliverable 0 document
>**1.3 System Users**
Chronos is built to aid many types of users of the Concordia community.
**1.3.1 Anonymous Users**
Anonymous users are users who chose not to log on to Chronos. The website will allow them to search for a class by course code and view the date, time, and term it is offered in as well as the professor teaching the course (if known) and the course description.
**1.3.2 Student Users**
Student users are defined as those users who log in to Chronos with their student ID. These users have access to the functionalities offered to anonymous users, but can also generate a customized schedule based on criteria established by the University and preferences input by the student.
**1.3.3 Professor Users**
Professor users are Concordia professors who log in to Chronos using their Concordia ID. Chronos will recognize these users as being professors and will allow them to view the classes they are teaching in a weekly calendar view.

I think the main area of confusion is where it says "and in the whole faculty as well" inside the project description but we can just scratch out professor users probably. By doing this though we would have to change the Deliverable 0 document contents.

**Comment by jhayarevalo** [comment]
Also I believe there is a missing association in your last comment @Shmeve (**Student** uses 1 **Schedule Generator**)

**Comment by Shmeve** [comment]
@jhayarevalo I agree with you on the missing relations and they have been added to the diagram and description. By faculty I think the project description meant the faculty of engineering, as in SOEN, COEN, MECH etc. I think we should cut professors out from our model and focus only on students users.

I have uploaded an updated version of deliverable0 to the google drive folder along with the latest version of the diagram (visio and png).
